### PR TITLE
Lower memory consumption

### PR DESCRIPTION
**Pull request by yglukhov**
1. Changed `getInstallFiles` -> `forEachInstallFile`.
2. Added `--disableValidation` option. Hidden for now.
3. Added git hash to `--version` output. Just in case...

### CODE REVIEW

**Comment by dom96** [comment]
@Araq suggested something like this:

```nim
proc findAux(dir: string, pkgInfo: PackageInfo; result: var seq[string]) = ...
proc findWithExt(dir: string, pkgInfo: PackageInfo): seq[string] = 
  result = @[]
  findAux(dir, pkgInfo, result)
```

I'm not sure if it will solve the memory issue though, any ideas?

**Comment by yglukhov** [comment]
This would definitely lower the memory pressure, but the point is, in case of verification we don't need the whole array of paths, as we process them one by one. So this solution is suboptimal.

File: src/nimble.nim [hunk 1]
```diff
@@ -445,7 +445,7 @@ proc installFromDir(dir: string, requestedVer: VersionRange, options: Options,
   createDir(pkgDestDir)
   # Copy this package's files based on the preferences specified in PkgInfo.
   var filesInstalled = initSet[string]()
-  for file in getInstallFiles(realDir, pkgInfo, options):
+  discard forEachInstallFile(realDir, pkgInfo, options) do(file: string) -> bool:
```
Comment (dom96) [inline src/nimble.nim]: Is the return `bool` value necessary? If not please remove it.

Rename to `iterInstallFiles`.

Nit: add space after `do`.

File: src/nimblepkg/packageinfo.nim [hunk 2]
```diff
@@ -430,30 +430,30 @@ proc checkInstallDir(pkgInfo: PackageInfo,
   if thisDir[0] == '.': result = true
   if thisDir == "nimcache": result = true
 
-proc findWithExt(dir: string, pkgInfo: PackageInfo): seq[string] =
-  ## Returns the filenames of the files that should be copied.
-  result = @[]
+proc forEachFileWithExt(dir: string, pkgInfo: PackageInfo,
```
Comment (dom96) [inline src/nimblepkg/packageinfo.nim]: Rename to `iterFilesWithExt`.

File: src/nimblepkg/packageinfo.nim [hunk 3]
```diff
@@ -430,30 +430,30 @@ proc checkInstallDir(pkgInfo: PackageInfo,
...
+    action: proc(f: string): bool): bool =
```
Comment (dom96) [inline src/nimblepkg/packageinfo.nim]: Nit: Space after `proc`.

Is the return value `bool` necessary?

File: src/nimblepkg/packageinfo.nim [hunk 4]
```diff
@@ -430,30 +430,30 @@ proc checkInstallDir(pkgInfo: PackageInfo,
...
     else:
       if path.splitFile.ext[1 .. ^1] in pkgInfo.installExt:
-        result.add path
+        if action(path): return true
 
-proc getFilesInDir(dir: string): seq[string] =
-  ## Returns a list of paths to files inside the specified directory and any
-  ## subdirectories that are in it.
-  result = @[]
+proc forEachFileInDir(dir: string, action: proc(f: string): bool): bool =
```
Comment (dom96) [inline src/nimblepkg/packageinfo.nim]: Rename to `iterFilesInDir`.

Remove return value if possible.

File: src/nimblepkg/packageinfo.nim [hunk 5]
```diff
@@ -430,30 +430,30 @@ proc checkInstallDir(pkgInfo: PackageInfo,
...
+      if forEachFileInDir(path, action): return true
     else:
-      result.add path
+      if action(path): return true
 
-proc getInstallFiles*(realDir: string, pkgInfo: PackageInfo,
-                      options: Options): seq[string] =
-  ## Returns a list of files within the ``realDir`` that should be installed.
-  result = @[]
+proc forEachInstallFile*(realDir: string, pkgInfo: PackageInfo,
```
Comment (dom96) [inline src/nimblepkg/packageinfo.nim]: Rename to `iterInstallFiles`

File: src/nimblepkg/packageinfo.nim [hunk 6]
```diff
@@ -430,30 +430,30 @@ proc checkInstallDir(pkgInfo: PackageInfo,
...
+                      options: Options, action: proc(f: string): bool): bool =
```
Comment (dom96) [inline src/nimblepkg/packageinfo.nim]: Remove return value if possible.

File: src/nimblepkg/packageparser.nim [hunk 7]
```diff
@@ -79,54 +79,53 @@ proc validatePackageStructure(pkgInfo: PackageInfo, options: Options) =
...
-                 "for source files, named '$3', but file '$1' " &
-                 "is in a directory named '$4' instead. " &
-                 "This will be an error in the future.") %
-              [file & ext, pkgInfo.name, correctDir, dir]
-          hint = ("If '$1' contains source files for building '$2', rename it " &
-                  "to '$3'. Otherwise, prevent its installation " &
-                  "by adding `skipDirs = @[\"$1\"]` to the .nimble file.") %
-              [dir, pkgInfo.name, correctDir]
-        raiseNewValidationError(msg, true, hint, true)
+    if ext == (ExtSep & "nim"):
```
Comment (dom96) [inline src/nimblepkg/packageparser.nim]: Don't indent this, instead I think you can use `return`.

**Comment by yglukhov** [comment]
Thanks for merging!

**Pull request merged**